src/Xamarin.Android.Build.Tasks/Utilities/XAJavaTypeScanner.cs

```diff
@@ -0,0 +1,167 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.IO;
+//using System.IO.MemoryMappedFiles;
+
+using Java.Interop.Tools.Cecil;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Mono.Cecil;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks;
+
+class JavaType
+{
+	public readonly TypeDefinition Type;
+	public readonly IDictionary<AndroidTargetArch, TypeDefinition>? PerAbiTypes;
+	public bool IsABiSpecific { get; }
+
+	public JavaType (TypeDefinition type, IDictionary<AndroidTargetArch, TypeDefinition>? perAbiTypes)
+	{
+		Type = type;
+		if (perAbiTypes != null) {
+			PerAbiTypes = new ReadOnlyDictionary<AndroidTargetArch, TypeDefinition> (perAbiTypes);
+			IsABiSpecific = perAbiTypes.Count > 1 || (perAbiTypes.Count == 1 && !perAbiTypes.ContainsKey (AndroidTargetArch.None));
```

Comment:
If there *is* an `AndroidTargetArch.None` value, perhaps `Type` should instead be a property?

```csharp
partial class JavaType {
    public TypeDefinition AbiNeutralType => PerAbiTypes.TryGetValue (AndroidTargetArch.None, out var t) ? t : throw new NotSupportedException (…);
}
```

Reply:
The existence of `AndroidTargetArch.None` implies that it *could* be possible for `perAbiTypes` to have `.None` + other ABIs, which feels like it should be not permitted.

Reply:
`AbiNeutralType` would have to always be returned, no exception thrown - it MUST be valid no matter what is in `PerAbiTypes`. `Type` described above saves one from having to write weird code to get **any** value from the `PerAbiTypes` like:

```csharp
TypeDefinition anyType;
foreach (var kvp in PerAbiTypes) {
   anyType = kvp.Value;
   break;
}
```

Which would be necessary if `Type` didn't exist.  As the code is implemented now, the readonly `Type` field is simply set whenever we add the first, any, TypeDefinition to the dictionary and that solves the problem in an elegant way.